### PR TITLE
expand home paths in fontconfig

### PR DESCRIPTION
**Pull request by snoyer**
Paths retrieved from `fontconfig` may contain `~` and need to be expanded.
There is probably a prettier/better way to do this but I've never looked at any Rust code before so I wouldn't know, feel free to treat this as a bug report rather that a PR and fix it the proper way if needed.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -337,12 +345,8 @@ impl Database {
 
             self.load_fonts_dir("/usr/share/fonts/");
             self.load_fonts_dir("/usr/local/share/fonts/");
-
-            if let Ok(ref home) = std::env::var("HOME") {
-                let home_path = std::path::Path::new(home);
-                self.load_fonts_dir(home_path.join(".fonts"));
-                self.load_fonts_dir(home_path.join(".local/share/fonts"));
-            }
+            self.load_fonts_dir(home_path.join(".fonts"));
```
Comment (RazrFalcon) [inline src/lib.rs:348]: I don't think this part should be changed. And you're potentially passing `~` here, which would have to be expanded.

File: src/lib.rs [hunk 2]
```diff
@@ -316,13 +316,21 @@ impl Database {
...
             #[cfg(feature = "fontconfig")]
             {
                 let mut fontconfig = fontconfig_parser::FontConfig::default();
                 match fontconfig.merge_config("/etc/fonts/fonts.conf") {
                     Ok(_) => {
                         for dir in fontconfig.dirs {
-                            self.load_fonts_dir(dir.path);
+                            let path = if dir.path.starts_with("~") {
+                                home_path.join(dir.path.strip_prefix("~").unwrap())
+                            }else{
```
Comment (RazrFalcon) [inline src/lib.rs:330]: Keep the spaces.

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 3]
```diff
@@ -316,13 +316,21 @@ impl Database {
         // Linux.
         #[cfg(all(unix, not(any(target_os = "macos", target_os = "android"))))]
         {
+            let home = std::env::var("HOME").unwrap_or("~".to_string());
```
Comment (RazrFalcon) [inline src/lib.rs:319]: This should be `.unwrap_or_else(|| "~".to_string());`.

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 3 above)
Comment (RazrFalcon) [reply src/lib.rs:319]: Or actually, we should not do this either. If `HOME` is not set - bail immediately.

**Comment by snoyer** [comment]
I'm not sure how much error handling and edge cases you want to cover.
Assuming doing the minimum of trying to expand `~/` and silently ignoring the case where `HOME` is not set (leaving `~/` in and letting `load_fonts_dir()` deal with non-existing path) maybe having a function is simpler?

```rust
            fn _expand_home<P: AsRef<std::path::Path>>(path: P) -> std::path::PathBuf {
                if let Ok(ref suffix) = path.as_ref().strip_prefix("~/") {
                    if let Ok(ref home) = std::env::var("HOME") {
                        let home_path = std::path::Path::new(home);
                        return home_path.join(suffix);
                    }
                }
                return path.as_ref().to_path_buf();
            }

            #[cfg(feature = "fontconfig")]
            {
                let mut fontconfig = fontconfig_parser::FontConfig::default();
                match fontconfig.merge_config("/etc/fonts/fonts.conf") {
                    Ok(_) => {
                        for dir in fontconfig.dirs {
                            self.load_fonts_dir(_expand_home(dir.path));
                        }

                        // Yes, stop here. No need to load fonts from hardcoded paths.
                        return;
                    }
                    Err(err) => {
                        // Fontconfig parsing failed? Try to load fonts from hardcoded paths then.
                        log::warn!("Failed to parse fontconfig because: {}", err);
                    }
                }
            }

            self.load_fonts_dir("/usr/share/fonts/");
            self.load_fonts_dir("/usr/local/share/fonts/");
            self.load_fonts_dir(_expand_home("~/.fonts"));
            self.load_fonts_dir(_expand_home("~/.local/share/fonts"));
```

That's pretty much the limit of my understanding of the language and of the codebase though :(

**Comment by RazrFalcon** [comment]
How about just:

```rust
                        for dir in fontconfig.dirs {
                            let path = if dir.path.starts_with("~") {
                                if let Ok(ref home) = std::env::var("HOME") {
                                    let home_path = std::path::Path::new(home);
                                    home_path.join(dir.path.strip_prefix("~").unwrap())
                                } else {
                                    continue;
                                }
                            } else {
                                dir.path
                            };
                            self.load_fonts_dir(path);
                        }
```

this is right after `match fontconfig.merge_config("/etc/fonts/fonts.conf") {`

This should be enough.

**Comment by RazrFalcon** [comment]
Done.

**Pull request closed without merging**